src/c++/library/common.h

```diff
@@ -330,13 +333,17 @@ class InferInput {
   /// \return Error object indicating success or failure.
   Error ByteSize(size_t* byte_size) const;
 
+  /// \return Whether the request should send binary input data.
+  bool BinaryInputData() const { return binary_input_data_; }
+
  private:
 #ifdef TRITON_INFERENCE_SERVER_CLIENT_CLASS
   friend class TRITON_INFERENCE_SERVER_CLIENT_CLASS;
 #endif
+  friend class HttpInferRequest;
```

Comment:
I was wondering  why do we need `friend` here?

Reply:
So that the member method [`HttpInferRequest::ConvertBinaryInputsToJSON()`](https://github.com/triton-inference-server/client/blob/matthewkotila-json-inputs-outputs/src/c%2B%2B/library/http_client.cc#L551) can call the private member method [`InferInput::GetNext()`](https://github.com/triton-inference-server/client/blob/main/src/c%2B%2B/library/common.h#L344).

When inputs are written to the request as binary, the binary buffers are accessed in [`InferenceServerHttpClient::PreRunProcessing()`](https://github.com/triton-inference-server/client/blob/matthewkotila-json-inputs-outputs/src/c%2B%2B/library/http_client.cc#L2059), which is the other friend declaration right above this.

To support json content type, it seemed easier to do the work of accessing the binary buffers from within the `HttpInferRequest` class, where the json is being written (and the binary buffers are converted to json to be written) rather than finding a way to do that logic in the `InferenceServerHttpClient` where the request json isn't really directly accessible.

Reply:
It seems like HTTP client can already call  `GetNext` and is a friend of the `InferInput`: https://github.com/triton-inference-server/client/blob/b66e4d4d084dd841f4199904dae95ecd61744687/src/c%2B%2B/library/http_client.cc#L308-L309